### ISSUE TITLE
How to pass request parameters as object ?

### ISSUE THREAD

**Issue by anton-arsentyev** [OpenFeign/feign#520]
Hello, is there a way to pass object as a parameter to interface and get all fields as query params ?

Example:

Interface:

```
@RequestLine("POST /api/doSomething")
void doSomething(@QueryObject Parameters parameters);
```

Model:

```
class Parameters {
  String name;
  String secondName;
}
```

And result request would be:

`{host}/api/doSomething?name=name123&secondName=sn222`

**Comment by imtiaz-m4** [comment]
I have found a solution here 
[Using request objects with Feign](http://source.coveo.com/2016/07/21/feign-request-object) by Jonathan Rochette and it looks pretty neat. Those extra annotation to correctly identify which parameter should go where.
But it will need extra coding in your application. Looks like Feign does not yet provide support for this.

I also need this solution built in Feign.

**Comment by DanielYWoo** [comment]
We want it! Let's say you have two mandatory parameters and an optional parameters, you will have to implement two methods: foo(m1, m2) and foo(m1, m2, o1). If you have more optional parameters you will have to implement a lot of overloaded methods, otherwise you can only implement a method with a lot of parameters and when it is called, you have to pass in a bunch of null values like foo(m1, m2, null, null, null, null, o5, null, null, o8) which is very ugly.

If we have object parameter we can do this: first, use a constructor to pass in all mandatory parameters, then use setters to set the optional parameters like this:

```
q = new FindUsersRequest(category, type);
q.setRowStart(10);
q.setGender(M);
feign.findUsers(q);

```

**Comment by smougenot** [comment]
Proposed solution is basic but fis the need in this PR (I also required it, so I share what I tried)
Hope it'll help

**Comment by kdavisk6** [comment]
This has been addressed by #667.  If there are additional use cases or use cases where this may not apply, please open a new issue.
